### PR TITLE
Add support to Qwen2-0.5B and Qwen2-1.5B.

### PR DESCRIPTION
**Pull request by fyabc**
#### Context

This PR:

- Add support to Qwen2-1.5B and Qwen2-0.5B, with full-parameter & lora finetune recipes.
- Change default finetune hyperparameters for Qwen2 recipes.
- Fix some bugs in tie_word_embeddings model loading & saving.
- Refactor `TiedEmbeddingTransformerDecoder` to align with the default `TransformerDecoder`.

Loss curves of Qwen2-0.5B and 1.5B full-parameter finetune on alpaca clean dataset have been uploaded to wandb (loss comparison curves are shown below)

What is the purpose of this PR? Is it to
- [x] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add unit tests for any new functionality
- [x] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
	- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### Loss Curves

All loss curves are smoothed with `smooth_ratio=0.6`.

loss-compare-0.5b-smooth0.6
wandb link: <https://api.wandb.ai/links/fyabc-123/jxbvzezh>

loss-compare-1.5b-smooth0.6
wandb link: <https://api.wandb.ai/links/fyabc-123/w2lqv1v6>

### CODE REVIEW

File: recipes/configs/qwen2/0.5B_full.yaml [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# Config for multi-device full finetuning in full_finetune_distributed.py
+# using a Qwen2 0.5B model
+#
```
Comment (SalmanMohammadi) [inline recipes/configs/qwen2/0.5B_full.yaml:3]: I feel a bit like we need to rethink `tune ls`. Here's the output just now:
```bash
RECIPE                                   CONFIG                                  
full_finetune_single_device              llama2/7B_full_low_memory               
                                         code_llama2/7B_full_low_memory          
                                         llama3/8B_full_single_device            
                                         llama3_1/8B_full_single_device          
                                         mistral/7B_full_low_memory              
                                         phi3/mini_full_low_memory               
                                         qwen2/7B_full_low_memory                
full_finetune_distributed                llama2/7B_full                          
                                         llama2/13B_full                         
                                         llama3/8B_full                          
                                         llama3_1/8B_full                        
                                         llama3/70B_full                         
                                         llama3_1/70B_full                       
                                         mistral/7B_full                         
                                         gemma/2B_full                           
                                         gemma/7B_full                           
                                         phi3/mini_full                          
                                         qwen2/7B_full                           
lora_finetune_single_device              llama2/7B_lora_single_device            
                                         llama2/7B_qlora_single_device           
                                         code_llama2/7B_lora_single_device       
                                         code_llama2/7B_qlora_single_device      
                                         llama3/8B_lora_single_device            
                                         llama3_1/8B_lora_single_device          
                                         llama3/8B_qlora_single_device           
                                         llama3_1/8B_qlora_single_device         
                                         llama2/13B_qlora_single_device          
                                         mistral/7B_lora_single_device           
                                         mistral/7B_qlora_single_device          
                                         gemma/2B_lora_single_device             
                                         gemma/2B_qlora_single_device            
                                         gemma/7B_lora_single_device             
                                         gemma/7B_qlora_single_device            
                                         phi3/mini_lora_single_device            
                                         phi3/mini_qlora_single_device           
                                         qwen2/7B_lora_single_device             
lora_dpo_single_device                   llama2/7B_lora_dpo_single_device        
lora_dpo_distributed                     llama2/7B_lora_dpo                      
lora_finetune_distributed                llama2/7B_lora                          
                                         llama2/13B_lora                         
                                         llama2/70B_lora                         
                                         llama3/70B_lora                         
                                         llama3_1/70B_lora                       
                                         llama3/8B_lora                          
                                         llama3_1/8B_lora                        
                                         mistral/7B_lora                         
                                         gemma/2B_lora                           
                                         gemma/7B_lora                           
                                         phi3/mini_lora                          
                                         qwen2/7B_lora                           
lora_finetune_fsdp2                      llama2/7B_lora                          
                                         llama2/13B_lora                         
                                         llama2/70B_lora                         
                                         llama2/7B_qlora                         
                                         llama2/70B_qlora                        
generate                                 generation                              
eleuther_eval                            eleuther_evaluation                     
quantize                                 quantization                            
qat_distributed                          llama2/7B_qat_full                      
                                         llama3/8B_qat_full    
```

Imo this isn't scaling well as we add more and more cool models and support configs for new techniques (imagine how many we'll have for multimodal!). It's getting unweildy. 

Not sure if this is already on your radar @joecummings @ebsmothers but I have an idea or two to address- happy to put an RFC?

File: torchtune/models/qwen2/_convert_weights.py [hunk 2]
```diff
@@ -68,7 +68,7 @@ def qwen2_hf_to_tune(
 
     for key, value in state_dict.items():
         if (
-            tie_word_embeddings and QWEN2_TIED_KEY not in key
+            tie_word_embeddings and QWEN2_TIED_KEY in key
```
Comment (SalmanMohammadi) [inline torchtune/models/qwen2/_convert_weights.py:71]: I missed this the first time, but is there a reason the checkpointing logic for tied embeddings in Qwen2 should be different from Gemma? Ref #1168, #1169

Comment (SalmanMohammadi) [reply torchtune/models/qwen2/_convert_weights.py:71]: @ebsmothers @joecummings

File: torchtune/models/qwen2/_model_builders.py [hunk 3]
```diff
@@ -40,6 +40,52 @@ def qwen2_7b() -> Union[TransformerDecoder, TiedEmbeddingTransformerDecoder]:
...
+        attn_dropout=0.0,
+        norm_eps=1e-06,
+        rope_base=1000000.0,
+        tie_word_embeddings=True,
+    )
+
+
+def qwen2_1_5b() -> Union[TransformerDecoder, TiedEmbeddingTransformerDecoder]:
+    """
+    Builder for creating a Qwen2 model initialized w/ the default 1.5B parameter values
```
Comment (SalmanMohammadi) [inline torchtune/models/qwen2/_model_builders.py:68]: Since `tied_embeddings=True`, we'll always be returning `TiedEmbeddingTransformerDecoder`, right?

File: torchtune/models/qwen2/_model_builders.py [hunk 4]
```diff
@@ -40,6 +40,52 @@ def qwen2_7b() -> Union[TransformerDecoder, TiedEmbeddingTransformerDecoder]:
     )
 
 
+def qwen2_0_5b() -> Union[TransformerDecoder, TiedEmbeddingTransformerDecoder]:
+    """
+    Builder for creating a Qwen2 model initialized w/ the default 0.5B parameter values
```
Comment (SalmanMohammadi) [inline torchtune/models/qwen2/_model_builders.py:45]: Could we please have a small comment here pointing out that `tied_embeddings` is enabled by default for 0.5B and 1.5B models? : )

File: recipes/configs/qwen2/0.5B_full_low_memory.yaml [hunk 5]
```diff
@@ -0,0 +1,77 @@
+# Config for single device full finetuning in full_finetune_single_device.py
```
Comment (ebsmothers) [inline recipes/configs/qwen2/0.5B_full_low_memory.yaml:1]: One thing that's a bit unclear (maybe this is our fault for how we've named our existing recipes), but configs like `*_low_memory` tend to use techniques that may be slower for the sake of being runnable on more constrained hardware. E.g. `llama2/7B_full_low_memory` should run in under 16GB of VRAM but won't be super fast. So I would question whether we actually need to apply these techniques for these smaller models: it may be better to use non-paged optimizer and disable optimizer in backward. (We could probably even get away with disabling activation checkpointing for these smaller models, but maybe that would bite us at longer sequence lengths)

File: recipes/configs/qwen2/0.5B_full.yaml (hunk 1 above)
Comment (joecummings) [reply recipes/configs/qwen2/0.5B_full.yaml:3]: Let's chat - this is definitely on my radar. Glad you caught it, too.

File: torchtune/models/qwen2/_convert_weights.py (hunk 2 above)
Comment (joecummings) [reply torchtune/models/qwen2/_convert_weights.py:71]: Good point, happy to have this filed as a refactor cleanup to start grouping some of these together in a "TIED_MODEL" key or something.

File: torchtune/models/qwen2/_model_builders.py [hunk 6]
```diff
@@ -111,3 +157,109 @@ def lora_qwen2_7b(
...
+
+
+def lora_qwen2_0_5b(
+        lora_attn_modules: List[LORA_ATTN_MODULES],
+        apply_lora_to_mlp: bool = False,
+        apply_lora_to_output: bool = False,
+        lora_rank: int = 8,
+        lora_alpha: float = 16,
+        lora_dropout: float = 0.05,
+        quantize_base: bool = False,
```
Comment (ebsmothers) [inline torchtune/models/qwen2/_model_builders.py:169]: nit: I think this is double-indented. Just kicked off CI so you can have linter signal but would also suggest running locally if you haven't already

**Review by ebsmothers — COMMENTED** [review]
Thanks for the PR! A few small comments here but this looks pretty good to me. Also the wandb link in the PR summary is broken for me, would definitely be interested to see the comparison with HF loss curves if that's possible

File: torchtune/models/qwen2/_convert_weights.py (hunk 2 above)
Comment (SalmanMohammadi) [reply torchtune/models/qwen2/_convert_weights.py:71]: Sorry if I'm missing something, but do we need a refactor here? I thought we ended up not needing any special checkpointing logic for Gemma, so it's just a matter of applying the same changes as in https://github.com/pytorch/torchtune/pull/1168/files - removing `qwen2/_convert_weights.py`, and remove any special logic in the checkpointers ([here](https://github.com/pytorch/torchtune/blob/72ef1ddac2d3d155b54e1c8323a6d865ba0bd0ee/torchtune/utils/_checkpointing/_checkpointer.py#L451), for example), so it goes through the default model save/load path.

**Comment by fyabc** [comment]
> Thanks for the PR! A few small comments here but this looks pretty good to me. Also the wandb link in the PR summary is broken for me, would definitely be interested to see the comparison with HF loss curves if that's possible

Hi, I have updated loss curves between torchtune and HF in description.

**Comment by ebsmothers** [comment]
> > Thanks for the PR! A few small comments here but this looks pretty good to me. Also the wandb link in the PR summary is broken for me, would definitely be interested to see the comparison with HF loss curves if that's possible
> 
> Hi, I have updated loss curves between torchtune and HF in description.

Thanks! Also bumping [this comment](https://github.com/pytorch/torchtune/pull/1247#discussion_r1698578721) @fyabc. Mainly I don't think we should need stuff like `PagedAdamW` or other memory optimizations that may slow down training for a 0.5B (or even 1.5B) model -- I imagine if we scrap these we can still support the same set of hardware but train faster. Let me know if this makes sense to you

File: recipes/configs/qwen2/0.5B_full_low_memory.yaml (hunk 5 above)
Comment (joecummings) [reply recipes/configs/qwen2/0.5B_full_low_memory.yaml:1]: I would say we delete these recipes entirely no? Or rename to `full_single_device` to denote that it doesn't have low memory optimizations?

**Comment by fyabc** [comment]
> > > Thanks for the PR! A few small comments here but this looks pretty good to me. Also the wandb link in the PR summary is broken for me, would definitely be interested to see the comparison with HF loss curves if that's possible
> > 
> > Hi, I have updated loss curves between torchtune and HF in description.
> 
> Thanks! Also bumping [this comment](https://github.com/pytorch/torchtune/pull/1247#discussion_r1698578721) @fyabc. Mainly I don't think we should need stuff like `PagedAdamW` or other memory optimizations that may slow down training for a 0.5B (or even 1.5B) model -- I imagine if we scrap these we can still support the same set of hardware but train faster. Let me know if this makes sense to you

Thank you for your suggestions! I will update the related 0.5B and 1.5B recipes.

File: recipes/configs/qwen2/0.5B_full_low_memory.yaml (hunk 5 above)
Comment (fyabc) [reply recipes/configs/qwen2/0.5B_full_low_memory.yaml]: Hi, I have renamed all `qwen*_full_low_memory` recipes to `qwen*_full_single_device`.

**Comment by fyabc** [comment]
> > Thank you for your suggestions! I will update the related 0.5B and 1.5B recipes.
> 
> Hi @fyabc! Anything we can do to help finish this up? We'd love to feature this on our README and post about it on our Discord channel, as well.

I have updated this PR to fix review comments. All `*_low_memory` recipes have been renamed to `*_single_device`, and model builder typing annotations have changed to more exact types.
I think this CR is ready to merge, and feel free for more suggestions.

**Review by joecummings — APPROVED** [review]
Awesome work as always :)
